### PR TITLE
[Routing] Generate URLs in compliance with PHP_QUERY_RFC3986

### PR DESCRIPTION
**Pull request by jameshalsall**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no? |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #19625 |
| License | MIT |
| Doc PR | reference to the documentation PR, if any |

### CODE REVIEW

File: src/Symfony/Component/Routing/Generator/UrlGenerator.php [hunk 1]
```diff
@@ -266,7 +266,13 @@ protected function doGenerate($variables, $defaults, $requirements, $tokens, $pa
         $fragment = isset($extra['_fragment']) ? $extra['_fragment'] : '';
         unset($extra['_fragment']);
 
-        if ($extra && $query = http_build_query($extra, '', '&')) {
+        if (defined('PHP_QUERY_RFC3986')) {
```
Comment (fabpot) [inline src/Symfony/Component/Routing/Generator/UrlGenerator.php]: In which PHP version was it added?

**Comment by fabpot** [comment]
Isn't it a bug fix instead of a new feature?

File: src/Symfony/Component/Routing/Generator/UrlGenerator.php (hunk 1 above)
Comment (jameshalsall) [inline src/Symfony/Component/Routing/Generator/UrlGenerator.php]: 5.4, the reason I've done it this way is to ease merging of upstream changes on 2.x branch

Comment (dunglas) [inline src/Symfony/Component/Routing/Generator/UrlGenerator.php]: 5.4 (the test can be removed in Symfony >= 3.0).

**Comment by jameshalsall** [comment]
I'm not so sure, it could have unintended side-effects depending on people's expectations of the `generate()` call.

File: src/Symfony/Component/Routing/Generator/UrlGenerator.php (hunk 1 above)
Comment (fabpot) [inline src/Symfony/Component/Routing/Generator/UrlGenerator.php]: The reason I asked was that you submitted the PR on master :) If this is a bug fix, this should be merged on 2.7 instead first.

Comment (fabpot) [inline src/Symfony/Component/Routing/Generator/UrlGenerator.php]: And if this is for master, then the constant is always defined.

Comment (jameshalsall) [inline src/Symfony/Component/Routing/Generator/UrlGenerator.php]: You're right :)

**Comment by jameshalsall** [comment]
Closed in favour of #19639 which is rebased on 2.7

**Pull request closed without merging**